### PR TITLE
Unit test get_df_from_iterator #373

### PR DESCRIPTION
**Pull request by christyanthony**
For SorrTask 373 - UNit_test_get_df_from_iterator
- Made a copy of iter_ so that it won't be consumed for function "get_df_from_iterator" in "helpers/hpandas.py"
- Added a unit test for function "get_df_from_iterator" in "helpers/test/test_hpandas.py" by creating 3 small dataframes

### CODE REVIEW

**Comment by christyanthony** [comment]
@DanilYachmenev ,

I have drafted the pull request as mentioned in the issue.

Also, I have added an assertion comparing length for the function `get_df_from_iterator`, but I am getting assertion error as the lengths are not matching. I have changed the code and did a unit test again, but this time even though the lengths were same i got the assertion error.

I am only facing the assertion error issue. Other than this, both the codes run perfectly. Please let me know your thoughts.

File: helpers/test/test_hpandas.py [hunk 1]
```diff
@@ -2992,3 +2992,58 @@ def test3(self) -> None:
...
+        """
+        Generate 3 small dataframes for the test.
+        """
+        # Define common columns.
+        columns = ["Column1", "Column2", "Column3"]
+        # Build dataframes.
+        idx1 = [0, 1]
+        data1 = [
+            ['A', 10, 1.1],
+            ['B', 20, 1.2],
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3015]: Lets do
```
d1 = {
    "num_col": [1, 2],
    "str_col": ["A", "B"],
}
d2 = {
    "num_col": [3, 4],
    "str_col": ["C", "D"],
}
d3 = {
    "num_col": [5, 6],
    "str_col": ["E", "F"],
}
df1 = pd.DataFrame(data=d1)
df2 = pd.DataFrame(data=d2)
df3 = pd.DataFrame(data=d3)
iter_ = iter([df1, df2, df3])
```

File: helpers/test/test_hpandas.py [hunk 2]
```diff
@@ -2992,3 +2992,58 @@ def test3(self) -> None:
...
+            ['F', 60, 1.6],
+        ]
+        df3 = pd.DataFrame(data3, columns=columns, index=idx3)
+        return df1, df2, df3
+
+    def test_get_df_from_iterator(self):
+        """
+        Unit test for get_df_from_iterator() function
+        """
+        # Get the test data
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3038]: see comments above about comments and empty spaces
but overall - no need to comment each step here

File: helpers/test/test_hpandas.py [hunk 3]
```diff
@@ -2992,3 +2992,58 @@ def test3(self) -> None:
...
+        """
+        Unit test for get_df_from_iterator() function
+        """
+        # Get the test data 
+        df1_in, df2_in, df3_in = self.get_test_data_for_iterator()
+        
+        # Create an iterator object with the test data
+        iterator = iter([df1_in, df2_in, df3_in])
+
+        # Call the get_df_from_iterator function
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3044]: do not reference code obj in comments 
https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#referring-to-an-object-in-code-comments

File: helpers/test/test_hpandas.py [hunk 4]
```diff
@@ -2992,3 +2992,58 @@ def test3(self) -> None:
...
+        result = hpandas.get_df_from_iterator(iterator)
+
+        # Compare outcomes using the signature approach
+        expected_result = pd.concat([df1_in, df2_in, df3_in]).sort_index()
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3048]: this is not a signature, this is code implementation
signature is a string
see `Test_compare_dfs` tests for reference

**Review by DanilYachmenev — COMMENTED** [review]
see the comments

**Comment by DanilYachmenev** [comment]
Also pls https://github.com/sorrentum/sorrentum/blob/master/docs/First_review_process.md#mention-the-issue

**Comment by christyanthony** [comment]
@DanilYachmenev ,

Thank You for the detailed review. 
I have made the requested changes based on your comments. Here are the updates that have been made:

- The TODO for function `get_df_from_iterator` has been reverted
- The code for the unit test has been re-written to include the implementation of the signature method and also modified the test dataframes.
- Docstrings, unnecessary comments, spaces, and lines have been revised and ensured to end with a period.

Please inform me if there are any other areas that require attention.

File: helpers/test/test_hpandas.py [hunk 5]
```diff
@@ -3057,3 +3057,40 @@ def test3(self) -> None:
...
+
+
+# #############################################################################
+
+
+class TestGetDfFromIterator(hunitest.TestCase):
+    """
+    Test concatenating dataframes from an Iterator into a single daatframe.
+    """
+    def test_get_df_from_iterator(self):
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3069]: -> `test1` since test class name reflects the same info

File: helpers/test/test_hpandas.py [hunk 6]
```diff
@@ -3057,3 +3057,40 @@ def test3(self) -> None:
...
         # Compare the actual outcome with expected one.
         self.check_string(act)
+
+
+# #############################################################################
+
+
+class TestGetDfFromIterator(hunitest.TestCase):
+    """
+    Test concatenating dataframes from an Iterator into a single daatframe.
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3067]: -> `Check that a dataframe is correctly build from an iterator of dataframes.`

and further on check grammar mistakes(like `daatframe`)
spell checker adds are good for this

File: helpers/test/test_hpandas.py (hunk 5 above)
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3069]: each function should have a return type hint 
`) -> None:`

File: helpers/test/test_hpandas.py [hunk 7]
```diff
@@ -3057,3 +3057,40 @@ def test3(self) -> None:
...
+    """
+    def test_get_df_from_iterator(self):
+        d1 = {
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3070]: Add comment
`# Build iterator of dataframes for the test.`

File: helpers/test/test_hpandas.py [hunk 8]
```diff
@@ -3057,3 +3057,40 @@ def test3(self) -> None:
...
+        }
+        d3 = {
+             "num_col": [5, 6],
+             "str_col": ["E", "F"],
+        }
+        df1 = pd.DataFrame(data=d1)
+        df2 = pd.DataFrame(data=d2)
+        df3 = pd.DataFrame(data=d3)
+        iter_ = iter([df1,df2,df3])
+        df_actual = hpandas.get_df_from_iterator(iter_)
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3086]: Add a comment above `# Run.`

File: helpers/test/test_hpandas.py [hunk 9]
```diff
@@ -3057,3 +3057,40 @@ def test3(self) -> None:
...
+        actual = hpandas.df_to_str(df_actual)
+        expected = """  num_col str_col
+        0        1       A
+        0        3       C
+        0        5       E
+        1        2       B
+        1        4       D
+        1        6       F
+        """
+        self.assert_equal(actual, expected, fuzzy_match = True)
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3096]: did you run linter?
kwargs inside functions are set without spaces between `=`
so `fuzzy_match=True`
linter is fixing it automatically

**Review by DanilYachmenev — COMMENTED** [review]
see the comments and pls post the test run logs (not all, just to confirm that it passes)

File: helpers/test/test_hpandas.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (christyanthony) [reply helpers/test/test_hpandas.py:3067]: @DanilYachmenev , are you asking me to remove the docstring under class and add the new comment under `test1` method?

File: helpers/test/test_hpandas.py [hunk 10]
```diff
@@ -3057,3 +3057,42 @@ def test3(self) -> None:
...
+        }
+        d3 = {
+             "num_col": [5, 6],
+             "str_col": ["E", "F"],
+        }
+        df1 = pd.DataFrame(data=d1)
+        df2 = pd.DataFrame(data=d2)
+        df3 = pd.DataFrame(data=d3)
+        # Run.
+        iter_ = iter([df1,df2,df3])
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3087]: did you run linter?
spacing here is broken, should be `iter([df1, df2, df3])`
could you run linter a check if it fixes it automatically?

File: helpers/test/test_hpandas.py [hunk 11]
```diff
@@ -3057,3 +3057,42 @@ def test3(self) -> None:
...
         self.check_string(act)
+
+
+# #############################################################################
+
+
+class Test_get_df_from_iterator(hunitest.TestCase):
+    def test1(self) -> None:
+        """
+        Check that a dataframe is correctly build from an iterator of dataframes.
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:3068]: -> built

File: helpers/test/test_hpandas.py [hunk 12]
```diff
@@ -3057,3 +3057,42 @@ def test3(self) -> None:
...
+        }
+        d2 = {
+            "num_col": [3, 4],
+            "str_col": ["C", "D"],
+        }
+        d3 = {
+             "num_col": [5, 6],
+             "str_col": ["E", "F"],
+        }
+        df1 = pd.DataFrame(data=d1)
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:3083]: better to keep 

```
df1 = ...
df1 = pd.DataFrame(data=df1)
```
for cohesiveness

Comment (christyanthony) [reply helpers/test/test_hpandas.py:3083]: sure

File: helpers/test/test_hpandas.py (hunk 11 above)
Comment (christyanthony) [reply helpers/test/test_hpandas.py:3068]: could you please elaborate on this. When I try to do "built" it is not recognizable.

Comment (DanilYachmenev) [reply helpers/test/test_hpandas.py:3068]: what do you mean @christyanthony?
in any case, `built` here is a grammatically correct spelling, right?
so do it this way

File: helpers/test/test_hpandas.py [hunk 13]
```diff
@@ -3057,3 +3057,42 @@ def test3(self) -> None:
...
+class Test_get_df_from_iterator(hunitest.TestCase):
+    def test1(self) -> None:
+        """
+        Check that a dataframe is correctly build from an iterator of dataframes.
+        """ 
+        # Build iterator of dataframes for the test.
+        df1 = {
+            "num_col": [1, 2],
+            "str_col": ["A", "B"],
+        }
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3074]: Gp meant to keep related code close to each other
so specify 1 df, then another one completely

and lets name dicts with data as `dataN`
```
data1 = {
    "num_col": [1, 2],
    "str_col": ["A", "B"],
}
df1 = pd.DataFrame(data=data1)
```

File: helpers/test/test_hpandas.py [hunk 14]
```diff
@@ -3057,3 +3057,42 @@ def test3(self) -> None:
...
+        df3 = {
+             "num_col": [5, 6],
+             "str_col": ["E", "F"],
+        }
+        df1 = pd.DataFrame(data=df1)
+        df2 = pd.DataFrame(data=df2)
+        df3 = pd.DataFrame(data=df3)
+        # Run.
+        iter_ = iter([df1, df2, df3])
```
[the remark this answers is not in the document]
Comment (christyanthony) [reply helpers/test/test_hpandas.py:3087]: I had to do it manually

File: helpers/test/test_hpandas.py (hunk 11 above)
Comment (christyanthony) [reply helpers/test/test_hpandas.py:3068]: @DanilYachmenev , I am not sure. I am getting name error. Is there a definition for `built` type in our codebase or is it defined in a separate module or package? Please let me know.

image

**Comment by christyanthony** [comment]
@DanilYachmenev , I didn't pull the request yet. But I have solved all the above conversations. Please reply the comment and will pull the request once `built` is sorted out.

File: helpers/test/test_hpandas.py (hunk 11 above)
Comment (samarth9008) [reply helpers/test/test_hpandas.py:3068]: You need to correct your comment grammatically.

`Check that a dataframe is correctly build from an iterator of dataframes.` -> `Check that a dataframe is correctly built from an iterator of dataframes.`

Comment (christyanthony) [reply helpers/test/test_hpandas.py:3068]: @samarth9008 @DanilYachmenev , you mean the comment. All this time I was thinking that it has something to do with return type. Thank you so much for the clarification.

**Comment by christyanthony** [comment]
> @DanilYachmenev , I didn't pull the request yet. But I have solved all the above conversations. Please reply the comment and will pull the request once `built` is sorted out.

It is sorted out.

**Review by DanilYachmenev — COMMENTED** [review]
lgtm

**Pull request merged**